packages/cli/src/commands/serve.ts

```diff
@@ -28,25 +29,38 @@ export default class Serve extends Command {
 
     await this.validateSpecPath(spec);
 
-    const components: TPrismHttpComponents<any> = !mock ? { config: { mock: false } } : {};
-    const server = createServer({ path: spec }, { components });
+    const server = this.createServer({ spec, mock });
     const address = await server.listen(port as number);
 
     this.log(address);
   }
 
-  private validateSpecPath(path?: string) {
-    if (!path) {
+  private createServer({ spec, mock }: { spec?: string; mock: boolean }) {
+    const components: TPrismHttpComponents<any> = {};
+    if (!mock) components.config = { mock: false };
+
+    if (spec && isHttp(spec)) {
+      return createServer(
+        { url: spec },
+        { components: { ...components, loader: httpLoaderInstance } }
+      );
+    } else {
+      return createServer({ path: spec }, { components });
+    }
+  }
+
+  private validateSpecPath(spec?: string) {
+    if (!spec || isHttp(spec)) {
       return;
     }
 
-    if (!existsSync(path)) {
```

Comment:
I wonder if the validateSpecPath is doing too much? It's being asked to know about HTTP, which means we are running the HTTP check twice, and we probably only need to run it once? 

Another _tiny_ thing but the existsSync and statSync are both a little bit race condition-prone anyway. Highly unlikely to be an issue, but a more robust approach is to attempt to read the file and handle the error if it does not work.

Maybe we can just do one of these and handle the rejections:

```
function readFileAsync(filename, encoding) {
    return new Promise((resolve, reject) => {
        fs.readFile(filename, encoding, (err, data) => {
            if (err) reject(err);
            else resolve(data);
        });
    });
}
```

This way we would not need a validateSpecPath method, but instead maybe a loadSpec method which returns the contents of the spec based on various inputs.

Reply:
1) Good spot on validateSpecPath. I refactored it to be more elegant.

2) Basically we cannot get rid of race condition here. Whatever we do here there will be still a read operation in graphite.

3) readFile (read-whole-file) is not a good approach for existence/dir checks in that particular situation. If we have big file / slow fs? Anyway it won't help use because of 2). It is also kind of temporary. I believe graphite should take care of loading http's.